File: src/year2022/day7/day7.py
#!/usr/bin/env python3

import sys
from collections import defaultdict
from pathlib import Path

filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"

path = Path("/")
dirs = set()
dir_size = defaultdict(lambda: 0)

for line in open(filename):
    line = line.strip()

    if line == "$ cd /":
        path = Path("/")
    elif line == "$ cd ..":
        path = path.parent
    elif line.startswith("$ cd "):
        assert "/" not in line[5:]
        assert len(line) > 5
        path /= line[5:]
    elif line == "$ ls":
        pass
    else:
        if line.startswith("dir "):
            pass
        else:
            size, file = line.split()
            size = int(size)
            dir_size[path] += size
    dirs.add(path)

total_dir_size = []
total = 0

# part one (and computing for part two)
part1 = 0
for dir in dirs:
    dir = dir.as_posix()
    s = 0
    for file_path, file_size in dir_size.items():
        file_path = file_path.as_posix()
        if file_path.startswith(dir):
            s += file_size
    total_dir_size.append(s)
    if s <= 100_000:
        part1 += s

    if dir == "/":
        total = s

print(part1)

# part two
for part2 in sorted(total_dir_size):
    if total - part2 + 30000000 <= 70000000:
        print(part2)
        break

File: src/year2022/day7/day7.rs
//! [Day 7: No Space Left On Device](https://adventofcode.com/2022/day/7)

use rustc_hash::FxHashMap;
use std::path::PathBuf;

struct Puzzle {
    total_dir_size: FxHashMap<String, usize>,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        let mut total_dir_size = FxHashMap::default();

        let lines = data.lines().collect::<Vec<_>>();

        let mut dir_size = FxHashMap::default();
        let mut current_path = PathBuf::from("/");

        dir_size.entry("/".to_string()).or_insert(0);

        for line in lines {
            if line == "$ cd /" {
                current_path = PathBuf::from("/");
            } else if line == "$ cd .." {
                assert_ne!(current_path.as_path().as_os_str().len(), 1);
                current_path.pop();
            } else if let Some(dir) = line.strip_prefix("$ cd ") {
                current_path.push(dir);

                let key = current_path.as_path().to_str().unwrap().to_string();
                dir_size.entry(key).or_insert(0);
            } else if line == "$ ls" || line.starts_with("dir ") {
                //  nothing to do
            } else if !line.is_empty() {
                // <size> <filename>
                let mut info = line.split(' ');
                let size = info.next().unwrap().parse::<usize>().unwrap();
                // let filename = info.next().unwrap();

                let key = current_path.as_path().to_str().unwrap().to_string();
                *dir_size.get_mut(&key).unwrap() += size;
            }
        }

        for dir in dir_size.keys() {
            let total = dir_size
                .iter()
                .map(|(x, size)| {
                    if dir.len() == 1
                        || x.starts_with(dir)
                            && (x.len() == dir.len() || {
                                x.chars().nth(dir.len()).unwrap() == std::path::MAIN_SEPARATOR
                            })
                    {
                        size
                    } else {
                        &0
                    }
                })
                .sum();

            total_dir_size.insert(dir.clone(), total);
        }

        Self { total_dir_size }
    }

    // Solves part one
    fn part1(&self) -> usize {
        let mut result = 0;
        for dir in &self.total_dir_size {
            if dir.1 <= &100_000 {
                result += dir.1;
            }
        }
        result
    }

    // Solve part two
    fn part2(&self) -> usize {
        let total = self.total_dir_size.get("/").unwrap();

        let mut sizes = self.total_dir_size.values().collect::<Vec<_>>();
        sizes.sort_unstable();

        for size in sizes {
            if total - size + 30_000_000 <= 70_000_000 {
                return *size;
            }
        }

        0
    }
}

#[must_use]
pub fn solve(data: &str) -> (usize, usize) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test01() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 95437);
        assert_eq!(puzzle.part2(), 24_933_642);
    }
}
